CONTRIBUTING.md (+README.md edits +CODE_OF_CONDUCT.md)

RangerTJ opened the pull request:
Goal: Addressing Issue #22 

Drafted a Contribution Guide with the intent of helping to consolidate links to existing documentation into a hub that has centralized guidelines / tips to get new contributors started. 

My goal was that, by the end of reading the guide, a new user would be able to:

- Find the resource page for what they want to contribute to
- Know where to find the current maintainers/lead developers and how to reach them (along with suggestions on how to helpfully stay engaged)
- Provide some "best practice" suggestions targeting newer folks that may be jumping into an open-source project for the first time
- Consolidate key development guide links on a single page that's easily referenced from the readme
- Address what to do if a security issue is encountered
- Heading-off potential common issues (like folks not noticing they're merging with the wrong repo)

While I _think_ I covered the essentials pretty thoroughly, based on scouring existing documentation and my chats with Mikou27, there may very well be gaps/some content issues that need to be addressed. If so, let me know, and I'll get them fixed up and update the PR ASAP! 

Notably, the style guide is a bit barebones, though I'm not sure if the project is actively looking to enforce any particular standard at the moment, which is why I mostly focused on C# "common standards" and high-level guidelines to promote code readability.

(Also, if this gets cross-threaded with the timing of any incoming updates, I can update my fork and let everyone know when the PR is ready to be reconsidered).

RangerTJ committed:
Final Draft v1

RangerTJ committed:
Update README.md

Added link from readme to contribution guide

RangerTJ committed:
Update README.md

Typos fixed

RangerTJ committed:
Update CONTRIBUTING.md

Rephrased and cleaned up donations section

RangerTJ committed:
Update CONTRIBUTING.md

a little more on bug guide

RangerTJ committed:
Update CONTRIBUTING.md

Eliminated ambiguous link description

RangerTJ committed:
Update CONTRIBUTING.md

RangerTJ committed:
Update CONTRIBUTING.md

made language more consistent on compiling guide link

RangerTJ commented:
Just caught a bug I'll need to fix - the readme links to the wrong repo's version of CONTRIBUTING.md.

RangerTJ committed:
Update README.md

Now links to the correct repo's version of the contributor guide

RangerTJ committed:
Update README.md

Testing relative path for guide link

RangerTJ committed:
Update README.md

attempt 2 at relative path

RangerTJ commented:
Swapped the readme link to a relative path to CONTRIBUTING.md, so it should be good to go now.

Snailedlt commented on README.md:
```diff
@@ -41,6 +41,8 @@ Note that Nucleus does not add multiplayer or co-op to single player games, the
 - Microsoft Visual C++ 2015-2019 Redistributable (both x86 and x64)
 
 # How can you contribute?
+If you would like to learn all about contributing to this project, whether bug reporting or working on the codebase, check out the [Nucleus Coop Contributing Guide](CONTRIBUTING.md)!
```
```suggestion
If you would like to learn all about contributing to this project, whether it's bug reporting or working on the codebase, check out the [Nucleus Coop Contributing Guide](CONTRIBUTING.md)!
```

RangerTJ replied:
Edit: I moved my comment from this reply to a new one for better visibility, since I resolved/implemented the readme text suggestion, but I think there's a couple other points I'd like to explore further to flesh out the document(s).

Snailedlt reviewed (commented):
This is good and all but it's a bit long. Lots of what is written here should be in a CODE_OF_CONDUCT.md file instead.

I'd wish there was more technical information, such as how to start the development server, dependencies, technologies used, etc...

Perhaps @Mikou27 could help with the technical part?

RangerTJ committed:
Update README.md

edit suggestion implemented

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ commented:
> This is good and all but it's a bit long. Lots of what is written here should be in a CODE_OF_CONDUCT.md file instead.
> 
> I'd wish there was more technical information, such as how to start the development server, dependencies, technologies used, etc...
> 
> Perhaps @Mikou27 could help with the technical part?

I'm moving my earlier reply to this comment for better visibility, since implementing the readme marked the conversation as resolved.

---

Thanks for the feedback!

I hadn't originally planned on tackling the Code of Conduct to try and limit scope, but... that said, I could definitely move the more "behavioral"/soft stuff to a code of conduct doc, if that'd be a helpful thing to add (though not sure if I should do that as a separate PR).

If I did do that, is there anything specific that you think should be moved to that doc / anything else that you think a code of conduct doc would need (other than the more traditional "don't be a jerk" type of stuff)? My initial thought is to move the whole "Communication" section there and then patch in the gaps for an initial Code of Conduct.

In terms of technical dependencies, were you thinking stuff like the version of .net that needs to be installed + a brief bit on what each of the submodule dependencies does? Or something a bit more detailed on them?

As I think about it, if the guide is starting to get a bit long too as it is, and it'd be good to keep it briefer, do you think it'd make sense to move relevant stuff to a Code of Conduct + maybe link to a more detailed article(s) about the dependencies and create that as a seperate issue to tackle? And then edit in a link to it from the contributor guide when that one is done? Depending on how much detail is needed, I could see that ballooning things quite a bit.

I wonder if a link to the detailed history doc from the SplitScreen.Me site might be helpful in here too? Maybe as a "read this first to know the development history of what you will be working on" kinda thing?

Thanks again! This is all really good stuff! Like I didn't even know there was a development server to work with, so that's exactly the kind of content gap that I was looking to identify.

Snailedlt commented:
@RangerTJ first of all, thank you for contributing and caring so much about getting it right, that is excellent!


Feel free to disagree or ask more about what I'm about to say :)

In my experience the CONTRIBUTING.md file usually contains a few sections:
- **How to contribute**
      - This just gives a brief overview of the different ways to contribute (you've already nailed this down in the PR, so good job :) )
- **How to report a Bug or suggest a feature**
      - Just a short explanation of how to contribute by opening an issue, discussion, thread on discord, etc... If issue templates are set up for the repo (which I believe is the case for us) they usually refer to those instead of writing out what info is needed in the markdown file. Something along the lines of "open an issue using the appropriate issue templates (link to issue templates)" should be okey. It would be good to mention here which things you should not open issues about (for example if you have issues with a specific handler, general controller issues, etc...)
- **How to develop locally**
      - This section is the bread and butter of the contributing file, and should be the most detailed part. This is mainly what I missed from your suggestion :) So what should be here? Everything needed to set up the project and start developing locally, how to open a PR, and the general workflow and guidelines. Workflow is the best practice and everything that's needed from the start of development to getting what you've developed out into production. This section is usually very heavy in code blocks with cli commands (for downloading dependencies, setup, building, starting the app, etc...), and code examples if needed. 
      - Feel free to break up this section into multiple sub sections, for example: **techstack**, **setup**, **local development/building**, **how to submit your code**, **how to make a release**
      - For all sections, but especially this one, the most important parts to explain are the project specific ones. If you feel like you need to explain anything that a seasoned developer should know already, it's often better to link to the appropriate documentation. For example it's good to explain how to install dependencies, but not so good to explain how you make a PR (this can instead be just a link to GitHub's documentation on how to make a PR.
- **How to submit your code / How to make a PR**
      - This typically explains the steps needed after you've implemented a feature, and want to submit it for code review, such as making sure tests pass locally (if the are any any), making a PR, etc...

As for the code of conduct, I'd recommend looking at the CODE_OF_CONDUCT.md file in this repo: https://github.com/devicons/devicon

We want it to be less general than that though. We especially want to make sure to include that we don't want to be associated with or included in illegal activities. For example we should include stuff like this in the file:
- "We do not make handlers that intentionally work for pirated/cracked games"
- "We do not endorse nor condone emulation, cracking and pirating games"


Hope that answered some of it. Feel free to ask more questions if you have any 😊

Mikou27 commented:
> > This is good and all but it's a bit long. Lots of what is written here should be in a CODE_OF_CONDUCT.md file instead.
> > I'd wish there was more technical information, such as how to start the development server, dependencies, technologies used, etc...
> > Perhaps @Mikou27 could help with the technical part?
> 
> I'm moving my earlier reply to this comment for better visibility, since implementing the readme marked the conversation as resolved.
> 
> Thanks for the feedback!
> 
> I hadn't originally planned on tackling the Code of Conduct to try and limit scope, but... that said, I could definitely move the more "behavioral"/soft stuff to a code of conduct doc, if that'd be a helpful thing to add (though not sure if I should do that as a separate PR).
> 
> If I did do that, is there anything specific that you think should be moved to that doc / anything else that you think a code of conduct doc would need (other than the more traditional "don't be a jerk" type of stuff)? My initial thought is to move the whole "Communication" section there and then patch in the gaps for an initial Code of Conduct.
> 
> In terms of technical dependencies, were you thinking stuff like the version of .net that needs to be installed + a brief bit on what each of the submodule dependencies does? Or something a bit more detailed on them?
> 
> As I think about it, if the guide is starting to get a bit long too as it is, and it'd be good to keep it briefer, do you think it'd make sense to move relevant stuff to a Code of Conduct + maybe link to a more detailed article(s) about the dependencies and create that as a seperate issue to tackle? And then edit in a link to it from the contributor guide when that one is done? Depending on how much detail is needed, I could see that ballooning things quite a bit.
> 
> I wonder if a link to the detailed history doc from the SplitScreen.Me site might be helpful in here too? Maybe as a "read this first to know the development history of what you will be working on" kinda thing?
> 
> Thanks again! This is all really good stuff! Like I didn't even know there was a development server to work with, so that's exactly the kind of content gap that I was looking to identify.

Hi man, imo you should continue on your path 
and improve over time (if you feel it) do it little but do it 😉.
Ideas and suggestions are good and all but don't get the job done. 
Do not hesitate to dm me if needed.

RangerTJ commented:
@Snailedlt  Thanks! That's super helpful!

Let me see if I can break down some of my thoughts and just make sure I'm understanding correctly, so I can create some more actionable chunks to work with.

>How to contribute

It sounds like I could largely leave this as-is, other than re-arranging some things and moving more "soft skill" chunks of it to the Code of Conduct.

>How to report a Bug or suggest a feature

Would you say that the current version of this is over-written, and could basically be simplified to "use the issue form and follow the template" with a link to the existing issue article in the FAQ (to keep it more concise), along with the explicit "don't make issues about these things" list? Maybe with a link to the code of conduct with a mention of contribution etiquette once the conduct doc exists?

>How to develop locally

I was originally thinking that this was handled by site [compiling guide](https://nucleus-coop.github.io/docs/compilation/) (since this had the git command to clone the repo with submodule dependencies), but it now occurs to me that compiling and setting up a dev environment are technically two different things, heh (even if compiling more or less requires setting up a dev environment). I was able to build successfully, so I initially assumed the compile guide might have been adequate, but... it sounds like it might not be, lol. Also, now that I look at it, that link mentions cloning the zerofox fork instead of that one, so that's another thing I'd need to address, even if it just stayed a link (unless we're supposed to clone from that fork for some reason?). 

Do you think it would make sense to go granular in terms of having setup instructions for a recommended IDE, or just focusing on the git commands (and stuff like .net devkit requirements) needed to successfully clone the repository so that it's ready for work?

I apologize for the quasi stream-of-consciousness here as I kinda outline how to tackle this in my head, lol.

Do you think something like the following would be sufficient for the "Local Development" section?

**Tech Stack**
- Clear listing of required programs and where to download them (C#/.net stuff, Visual Studio, etc.)
- OS requirements, versions, etc.
- Subsection: Dependent Submodules (brief explanations of what each does within Nucleus Coop and links to their respective repositories) - if I understand correctly, it sounds like we may not go into deep detail on them here, since if folks want to work on the submodules, they can do that on their respective repos.
- Now that I think about it, might also be helpful to have a brief summary of how these interdependent modules interact.

**Local Setup**
- Make sure development dependencies installed (see tech stack)
- Brief forking instruction with link to github article
- Brief branch guide here (emphasizing that "master" is the primary working branch)
- Git command to clone this repo with the submodule dependencies (link to github clone
- Building ProtoInput Instructions (since it's a dependency that requires some setup)
- (I think, other than ProtoInput needing to be built, all other submodule dependencies should be taken care of by the recrusive cloning of the repo?)
- [At this point I _think_ you should have a working development environment? Unless I'm missing something?]

**Building Nucleus Coop**
- Description of the different build options (Release, Debug, DebugTest, etc. - I think I need to check with Micou27 about these)
- Build instructions from the compiling guide - not sure if I should just link to the compiling guide here (since the clone link on it may be out of date?), or reproduce it in the guide doc/rephrase it?
- Running your build to test changes (out of curiosity, is there a list anywhere of what all the non NucleusCoop.exe executables are for/do?)

**How to Submit**
- Reminder to thoroughly test changes before you submit (I feel like this could, and probably should be a very deep rabbit hole at some point, but it's probably a whole issue in itself). Not sure if there's any existing testing tools I may have missed?
- "Standard" content replaced with a link to the github article on Pull Requests.
- Note about verifying your PR merge target branch (slim down the current super-detailed version)
- (Side note: Does everyone else default to merging with zerofox's fork too? I'm not sure if that's just a "me" thing or a universal issue/pitfall that's worth calling out, like I did in the current draft).

**Release Builds**
- Release criteria
- Who determines release readiness
- Who creates it
- Release schedule/frequency (not sure if this is a set thing, so maybe not this part?)

RangerTJ committed:
Update CONTRIBUTING.md

First pass at expanding setup part

RangerTJ commented:
Just finished a first pass at updating the doc by expanding the setup guide. Note that I specifically have _not_ yet given the document a simplification pass (so as of right now, it's _even longer_, lol), so expect trimming of some of the stuff that could be explained by Github articles or could be passed along to existing SplitScreen.Me ones (such as the bug report guide), along with a bunch of stuff moving to an "etiquette tips" part of a Code of Conduct doc. Also added a bit about NO PIRATES to the communication section to make sure I don't forget that, lol.

I think there might be some technical issues with some of the setup instructions, but those are things that I think I'll need to DM Mikou27 about. Theoretically they should be good though (assuming the other contribution guide is accurate), there just may be some missing bits to sort out, like compatibility with VS2022, etc. So I think it _should_ be more about adding a little extra version requirement info vs. totally re-writing anything.

RangerTJ committed:
first draft of the code of conduct

Possibly need to revise part about who to contact about behavioral issues.

RangerTJ commented:
I went ahead and threw together a 1st draft of a code of conduct based on the template you linked. For now it's largely the same, though I added an explicitly-headed section that put the anti-piracy stance front and center, along with migrating some of the etiquette points I'd drafted earlier from CONTRIBUTING.md (though some of those may still need to be shrunk a bit). I also added a link to the Discord channel rules... Which I could probably put in the doc itself, if that made more sense (though not sure if the Discord rules are just specific to discord or not).

There's currently a potential placeholder in the portion that covers who to report behavior violations to. In the template/devicons example it looks like there's a dedicated "reporting stuff" email, but I'm not sure if Nucleus Coop has the same? For now I just said to go to the support channel and ping the mods/admins as a stopgap measure, but not entirely sure if there's a preferred approach (or if it should go into more detail if that stays the approach, like ping to mention here's an issue, then mods will ask for details in DM's to protect privacy of reporting or some such... this would also probably apply to the security issues part of CONTRIBUTING.md, I'd imagine).

Note that I haven't yet culled the stuff I moved over from CONTRIBUTING.md yet.

RangerTJ committed:
Edits to COC + Removed COC stuff from CONT md file

Changed security issue to contact via pings and then follow-up instead of DM'ing people directly. "The Process" section could probably be deleted if it's deemed redundant with the detailed part that follows. Left it in for now as a "what to expect" outline/reference.

RangerTJ commented:
Trimmed the stuff I moved to the code of conduct from CONTRIBUTING.md + some minor code of conduct edits. Added a bit to code of conduct about avoiding unsolicited DM's, and the best way to reach people until you are asked/given the OK to talk to someone in DM's.

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,203 @@
+If you’d like to expand the list of supported games (or improve an existing game’s support), check out the detailed [Handler Development Guide](https://www.splitscreen.me/docs/create-handlers)!
+
+## Issue Contribution
+If you encounter an issue while using Nucleus Coop, or have an idea for a new feature, you can create a [GitHub “Issue”](https://github.com/SplitScreen-Me/splitscreenme-nucleus/issues) using the “New Issue” button on the Issue page. This will take you to a selection page where you can choose the most relevant template for your issue.
```
```suggestion
If you encounter an issue while using Nucleus Coop, or have an idea for a new feature, you can [create an issue](https://github.com/SplitScreen-Me/splitscreenme-nucleus/issues/new/choose). This will take you to a selection page where you can choose the most relevant template for your issue.
If you want to suggest a new feature or report a bug for any of our other projects (for example our discord bot, or websites), please check out the appropriate repositories: https://github.com/orgs/SplitScreen-Me/repositories
```

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,203 @@
+## Issue Contribution
+If you encounter an issue while using Nucleus Coop, or have an idea for a new feature, you can create a [GitHub “Issue”](https://github.com/SplitScreen-Me/splitscreenme-nucleus/issues) using the “New Issue” button on the Issue page. This will take you to a selection page where you can choose the most relevant template for your issue.
+
+The SplitScreen.Me FAQ has a great article that discusses [Bug Reporting](https://www.splitscreen.me/docs/faq/#18--where-can-i-report-a-bugissue) in more detail, including how to enable and view log files.
```
very good! 💯

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,203 @@
+## Security Note
+If you identify a security issue in the code of Nucleus Coop, please do not post a public issue identifying the security threat. This potentially highlights the issue for malicious users to take advantage of the vulnerability. 
+
+Instead, please ping @Administrators, @Moderators, or @Developers in the support channel of the Nucleus Coop Discord server, and only share the details once a secure channel of communication has been established.
```
```suggestion
If you identify a security issue in the code of Nucleus Coop, please do not post a public issue identifying the security threat. This potentially highlights the issue for malicious users to take advantage of the vulnerability. 
Instead, please ping @Administrators, @Moderators, or @Developers in the support channel of the Nucleus Coop [Discord server](https://discord.com/invite/a9ssM5pxTW), and only share the details once a secure channel of communication has been established.

Alternatively you can [create a repository security advisory](https://docs.github.com/en/code-security/security-advisories/working-with-repository-security-advisories/creating-a-repository-security-advisory)
```

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,203 @@
+
+5.	**Done!** Once your code is reviewed and accepted, it will be merged into the codebase of the project! Congratulations! You should see your changes in the next release build!
+
+To learn more about [forking](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/fork-a-repo), [cloning](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository), and [pull requests](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a ...
```
Very good note! Just making it a bit more cohesive
```suggestion
Check out the official GitHub articles to learn more about [forking](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/fork-a-repo), [cloning](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository), and [pull requests](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request).
```

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,203 @@
+
+    Close the Proto Input project out when you're done.
+
+You are now free to work on the Nucleus Coop source code - all dependencies in place! If you get any errors in compiling along the way, you may need to install and run project solution files from an earlier version of Visual Studio or the .NET SDK for better project compatibility.
```
```suggestion
1. Fork the repo
2. clone the repo
     `git clone https://github.com/JoeyJoeJoeShabadooFork/splitscreenme-nucleus`
3. Navigate to the root folder and update the submodules to make sure that all submodule dependencies are downloaded and up-to-date by running the following command:
     ```sh
     cd splitscreenme-nucleus
     git submodule update --init --recursive
     ```
4. Batch Build **Proto Input** by opening its solution file in Visual Studio and batch building it ("Submodules\ProtoInput\src\ProtoInput\ProtoInput.sln" within the project root). 

    The cleanest way is probably to sort by "Solution Config" header under the Batch Build menu and check the boxes that correspond with the version you would like to compile (make sure to check both x86 and x64 boxes for your selected build type).

    Note: you may also refer to the [Compilation](https://nucleus-coop.github.io/docs/compilation/) article for a more detailed example (with screenshots) of creating builds from the source code.

    Close the Proto Input project out when you're done.

You are now free to work on the Nucleus Coop source code - all dependencies in place! If you get any errors in compiling along the way, you may need to install and run project solution files from an earlier version of Visual Studio or the .NET SDK for better project compatibility.
```

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,203 @@
+
+When you are ready to test any changes you make, compile **Nucleus Coop** by opening its solution file and running a batch build on it within Visual Studio ("Master\NucleusCoop.sln"). Similar to Proto Input, on the Batch Build menu, select all the x86 and x64 options for the build you would like to compile. 
+
+Once you have selected your build options and run the batch build command, your output build directory should now have a fully functioning versio of the program and everything it needs to run! To test your changes live within Nucleus Coop itself, go to your build directory and run "NucleusCoop.exe" to test out your new build.
```
```suggestion
Once you have selected your build options and ran the batch build command, your output build directory should have a fully functioning version of the program and everything it needs to run! You can now go to the build directory and run "NucleusCoop.exe" to test out your new build.
```

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,203 @@
+
+Once you have selected your build options and run the batch build command, your output build directory should now have a fully functioning versio of the program and everything it needs to run! To test your changes live within Nucleus Coop itself, go to your build directory and run "NucleusCoop.exe" to test out your new build.
+
+As a reminder, please take care to follow the Coding Standards for the project when adding any code-based contributions.
```
Link to code standards
```suggestion
As a reminder, please take care to follow the [Code Standards](#code-standards) for the project when adding any code-based contributions.
```

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,203 @@
+
+As a reminder, please take care to follow the Coding Standards for the project when adding any code-based contributions.
+
+Test your changes **thoroughly** before submitting a Pull Request. Whenever approrpriate, it is strongly recomended that you develop unit-tests with a good test spread to make sure that your new features are working as intended and are capable of handling anticipated edge cases. In cases where this is not possible, please be sure to follow a similar testing paradigm where both average and edge us ...
```
```suggestion
Test your changes **thoroughly** before submitting a Pull Request. Whenever appropriate, it is strongly recommended that you develop unit-tests with a good test spread to make sure that your new features are working as intended and are capable of handling anticipated edge cases. In cases where this is not possible, please be sure to follow a similar testing paradigm where both average and edge use cases are tested. In the long run, this will save the project a lot of time, avoiding tedious back-tracking and bug-smashing by proactively ensuring that easily-preventable bugs don't sneak into the master build.
```

Snailedlt commented on README.md:
Hmm, perhaps everything under `How can you contribute?` can be removed or moved to CONTRIBUTING.md? Also the heading should probably be changed to `How can I contribute?`,  `How to contribute` or `Contributing`

RangerTJ replied:
Makes sense to me - I think everything but mention of the Splitscreen Dreams discord is covered in the new contributing doc (and that can be rectified easily). Though I can't actually find a link to it anywhere, heh. Not sure if that reference is deprecated? Or maybe things have migrated to mostly be on the Nucleus Coop server now? 

Edit: Looks like it's invite-only (from a quick search of Nucleus Coop discord), so not sure if it's better to advertise it here, or have folks get pointed towards it incidentally while they work on contributing if it comes up organically during collaboration?

I've incorporated some of the lines / similar phrasing in the contributing doc, where I felt it would add to the contributing guide a bit, in my next commit. When I first stumbled on this project, I really liked how welcoming and enthusiastic the messaging was, do I definitely want to make sure that aspect is preserved with any changes.

Snailedlt replied:
> Looks like it's invite-only (from a quick search of Nucleus Coop discord)

Correct! It's invite only by preference of the admins in that discord server, so we shouldn't link to it here. If people want to get invited to Splitscreen Dreams they can ask in the Nucleus Coop discord server, but usually it happens through collaboration like you say :)

> I really liked how welcoming and enthusiastic the messaging was, do I definitely want to make sure that aspect is preserved with any changes.

That's good to hear! A lot of people has worked on the README, so it definitely has personality ;)

RangerTJ replied:
Gotcha, I went ahead and committed the readme change!

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,203 @@
+
+**Make sure that the _base_ repository is set to "/SplitScreen-Me/splitscreenme-nucleus"!** 
+
+Otherwise, you will be attempting to merge your changes with a different repository. Double-check that you will be merging with the correct head branch as well (generally "master") Also verify that the _head_ repository is set to your fork and the branch of your fork where you made all the changes that you would like to implement.
```
```suggestion
Otherwise, you will be attempting to merge your changes with a different repository. Double-check that you will be merging with the correct _base_ branch as well (generally "master") Also verify that the _head_ repository is set to your fork and the branch of your fork where you made all the changes that you would like to implement.
```

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,203 @@
+* Any questions or other comments for contribution reviewers
+
+#### Review Process
+All pull requests will approved by a sufficient number of project maintainers (those with write access to the project repository) before any changes are merged. Before aproving changes, reviewers will determine whether:
```
I would simplify this a bit. If they've come this far they probably know what a maintainer is :)
```suggestion
All pull requests will need to be approved by a sufficient number of maintainers before any changes are merged. Before approving changes, reviewers will make sure that:
```

RangerTJ replied:
I'm incorporating this suggestion manually, since it needs to go along with a slight tweak to phrasing of one or two of the following bullet points - will be implemented in the next uploaded commit.

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,203 @@
+* Sufficient testing occurred
+* Quality standards are upheld
+
+While it is certainly _possible_ that a quality pull request might be accepted right away, it should be expected that there will be some changes requested of the pull request, and issues that need to be discussed and resolved before it is approved. The pull request post itself is a great place to have these conversations, though contributors should not be afraid to directly message fellow develop ...
```
```suggestion
While it is certainly _possible_ that a high quality pull request might be accepted right away,chagne requests, questions and discussions should be expected before the pull request can be merged. Discussions around the pull request should preferably be held in the pull request comments so it's easy for multiple people to follow the progress... but don't hesitate to reach out directly to the developers and maintainers either.
```

RangerTJ replied:
Incorporating this one manually, since I have a couple proposed (very slight) tweaks for the suggestion - should be in the next commit of CONTRIBUTING.md.

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,203 @@
+### Implementing Changes
+
+#### Approval
+Once the review process is completed to the satisfaction of project enough project maintainers, a project maintainer will merge the pull request with the "master" branch, incorporating any changes within the pull request.
```
```suggestion
Once the review process is completed to the satisfaction of enough maintainers, a maintainer will merge the pull request with the "master" branch, incorporating any changes within the pull request.
```

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,203 @@
+#### Approval
+Once the review process is completed to the satisfaction of project enough project maintainers, a project maintainer will merge the pull request with the "master" branch, incorporating any changes within the pull request.
+
+At this point, the contributor's job is done! If your pull request made it this far, congratulations! Your contribution is now part of the project!
```
```suggestion
At this point, the contributor's job is done! If your pull request made it this far, congratulations! Your contribution is now part of the project!
Thank you for contributing!
```

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,203 @@
+At this point, the contributor's job is done! If your pull request made it this far, congratulations! Your contribution is now part of the project!
+
+#### Release Builds
+Periodically, the current lead developer(s) of Nucleus Coop will decide when enough changes have been incorporated into the master branch to justify a new release. After sufficient testing by project maintainers to ensure build stability, a release will be created and published alongside a change log of implemented changed as improvements. Releases include source code along with a more convenient ...
```
```suggestion
Periodically, the current lead developer(s) of Nucleus Coop will decide when enough changes have been incorporated into the master branch to justify a new release. To ensure build stability, a release will be created and published alongside a change log of implemented changes after sufficient testing by maintainers. Releases include source code along with a more convenient installer and a "portable" (pre-compiled zip folder) option for users to run the program.
```

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,203 @@
+Currently, there is no set release schedule, and releases will be published when ready and verified stable.
+
+### Code Standards
+* When in doubt, structure your code to err on the side of readability!
```
As a non-native English speaker, I had trouble understanding what err meant in this sentence. Perhaps it should be re-worded? Feel free to disagree
```suggestion
* When in doubt, structure your code to prioritize readability!
```

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,203 @@
+        
+    [submodule "Submodules/TypescriptSyntaxPaste"]
+        path = Submodules/TypescriptSyntaxPaste
+        url = https://github.com/nhabuiduc/TypescriptSyntaxPaste.git
```
This can be moved into a code block, like so: 

```ini
    [submodule "Submodules/x360ce"]
	    path = Submodules/x360ce
	    url = https://github.com/lucasassislar/x360ce

    [submodule "Submodules/ProtoInput"]
        path = Submodules/ProtoInput
        url = https://github.com/ilyaki/protoinput

    [submodule "Submodules/nukeupdater"]
        path = Submodules/nukeupdater
        url = https://github.com/lucasassislar/nukeupdater
        
    [submodule "Submodules/TypescriptSyntaxPaste"]
        path = Submodules/TypescriptSyntaxPaste
        url = https://github.com/nhabuiduc/TypescriptSyntaxPaste.git
```

RangerTJ replied:
I fought with the .md formatting a lot with this part, lol. The current version on my fork actually displays similar to that, oddly enough (https://github.com/RangerTJ/splitscreenme-nucleus/blob/master/CONTRIBUTING.md). I'm not entirely sure _why_ though. It might be displaying as another block type and not a code block, since I didn't use the '>' lead in for it. 

I noticed the one on my fork has sharp edges vs. rounded edges on the block). I tried using a leading '>' earlier but it kept giving me really weird formatting issues, where the brackets would be outside the block but the indented code chunks would be contained within it (or depending on how I spaced things, it'd wrap the text vs. letting me indent/create new lines). I'll play with it some more and see if I can get it to behave, heh.

Edit: Copilot says bracketing the text in triple backticks may do the trick, so I'll give that a shot. 
Edit 2: Yup, that seemed to do the trick!

RangerTJ replied:
Should now be implemented!

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,203 @@
+
+**Compare**
+>compare: YourSourceBranch
+
```
This section could probably be shortened quite a bit, though I'm not sure exactly how 😅

RangerTJ replied:
Yeah, I was kinda iffy on leaving this one as-is, since it's kinda long, lol. If we're operating on the assumption that the contributor knows how to use Github and do a Pull Request, we could probably shave off all the detailed parts and just leave the comments highlighting that it might default the base to an older fork of the project (and include the reminder about changing the head first to avoid having to do it all again).

RangerTJ replied:
New commits should have a heavily simplified version of this section. May still be a bit of redundancy with the Pull Request part of the article. I'm not _entirely_ sure it's absolutely necessary, since anyone who knows how to use GitHub should probably be able to notice that when the head/base are wrong, lol.

Maybe just take everything from (in the new version) "GitHub may default..." on down and put it in the Pull Request section instead? And just leave this section for a quick note/clarification on the project branches (giving a space to edit in details later in case the current "master" paradigm ever changes).

RangerTJ replied:
Implemented a first pass of a hatchet-job to this section, but may want to follow-up on my previous comment and further trim it/restructure it, depending on what everyone thinks.

Snailedlt replied:
Imo it looks okey now. I'll resolve this and make another suggestion related to this in a new review :)

Snailedlt reviewed (changes requested):
This already looks much better!
I didn't look through the code of conduct yet, but I thought I'd just give some quick feedback on Contributing.md asap, so you don't have to wait too long :)
Overall very good, no major changes needed imo, just a few minor ones

RangerTJ committed:
Update CONTRIBUTING.md

incorporating SnaledIt's suggestion

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Update CONTRIBUTING.md

incorporating SnaledIt's suggestion / the reference to creating a repository security advisory

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Update CONTRIBUTING.md

incorporating SnaledIt's suggestion

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Update CONTRIBUTING.md

Fixed readability of the code readability line.

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Update CONTRIBUTING.md

Thank you for the suggestion!

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ commented:
Just wanted to add a quick note - I chatted with Mikou27 last night/this morning and figured out my issue with compiling ProtoInput was a missing dependency for it. I'll see if I can track down what it is and add it to the pre-req part of the doc.

RangerTJ committed:
Update CONTRIBUTING.md

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Update CONTRIBUTING.md

Much more concise!

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Code block cleanup and branch tips trimming

RangerTJ committed:
Proposed minor style edits to setup section

RangerTJ committed:
Update CONTRIBUTING.md

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Update CONTRIBUTING.md

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Apply suggestions from code review

suggestions implemented

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Update CONTRIBUTING.md

Thanks for catching that typo!

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Additional edits / manual suggestion implents

RangerTJ committed:
Merge pull request #3 from SplitScreen-Me/master

Merge pull request #74 from RangerTJ/master

RangerTJ committed:
Phrasing edit

Slightly tweaked phrasing for Splitscreen Dreams reference incorporated in contributuing.md from the original readme text (which may be trimmed down substantially).

RangerTJ commented:
Finished incorporating most of the recent suggestions (and also synch'd my fork with the master repo, since it looks like it had been 1 commit behind... ironically due to the meta form PR I submitted, lol).

I think there's a few notable points to work on more before I think this PR is fully ready for a "final" review:
- Finalizing README.md content (if we're ok just trimming it to the link sentence + the "Thank You" part, that'd make sense to me)
- Phrasing for Splitscreen Dreams server references in CONTRIBUTING.md (assuming it's moved/removed from the readme)
- Finalizing content/placement of content regarding merging issues with the default branches
- I still need to add a reference to the missing dependency to make sure compiling ProtoInput works (I've spent most of today on implementing doc changes so far, but I'm _hoping_ this part doesn't take too long, lol - I'll poke around the ProtoInput page a bit and see if I can get it sorted so that it'll compile with no errors). I could also add a bit about how you can just drop-in the proto input files from the latest release build as a work-around in the troubleshooting section (in case people are getting really stuck on submodule compiling).

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -113,84 +117,71 @@ It is a good idea to create a detailed (but concise) comment to accompany the in
 * Quality standards are upheld
 
-While it is certainly _possible_ that a quality pull request might be accepted right away, it should be expected that there will be some changes requested of the pull request, and issues that need to be discussed and resolved before it is approved. The pull request post itself is a great place to have these conversations, though contributors should not be afraid to directly message fellow develop ...
+While it is certainly _possible_ that a quality pull request might be accepted right away, change requests, questions, and discussions should be expected before the pull request can be merged. Discussions around the pull request should preferable be held in the pull request comments. This way it is easy for multiple people to follow progress... but don't hesitate to reach out directly to the deve ...
```
Whooops, looks like my previous suggestion might have had a typo
```suggestion
While it is certainly _possible_ that a quality pull request might be accepted right away, change requests, questions, and discussions should be expected before the pull request can be merged. Discussions around the pull request should preferably be held in the pull request comments. This way it is easy for multiple people to follow progress... but don't hesitate to reach out directly to the developers and maintainers either!
```

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -113,84 +117,71 @@ It is a good idea to create a detailed (but concise) comment to accompany the in
+While it is certainly _possible_ that a quality pull request might be accepted right away, change requests, questions, and discussions should be expected before the pull request can be merged. Discussions around the pull request should preferable be held in the pull request comments. This way it is easy for multiple people to follow progress... but don't hesitate to reach out directly to the deve ...
+
+
+it should be expected that there will be some changes requested of the pull request, and issues that need to be discussed and resolved before it is approved. The pull request post itself is a great place to have these conversations, though contributors should not be afraid to directly message fellow developers either, when doing so is helpful to resolving any issues.
```
Isn't this just saying the same as what's written above?
```suggestion
```

RangerTJ replied:
Yup! Looks like the old phrasing (or maybe an edit in progress I forgot to delete) somehow survived. Should be dead after this commit though. For real this time!

Snailedlt commented on CONTRIBUTING.md:
````diff
@@ -113,84 +117,71 @@ It is a good idea to create a detailed (but concise) comment to accompany the in
-        path = Submodules/TypescriptSyntaxPaste
-        url = https://github.com/nhabuiduc/TypescriptSyntaxPaste.git
+Even if some of these dependencies are not in active use, the git "update" command may fail to execute if entries and URLs for **all submodule** dependencies cloned from your fork are not found in the .gitmodules file. Add anything missing, save the changes to .gitmodules and the command should now work.
+```
````
to get code highlighting :)
```suggestion
```ini
```

It should then look like this:
```ini
[submodule "Submodules/x360ce"]
    path = Submodules/x360ce
    url = https://github.com/lucasassislar/x360ce
[submodule "Submodules/ProtoInput"]
    path = Submodules/ProtoInput
    url = https://github.com/ilyaki/protoinput
[submodule "Submodules/nukeupdater"]
    path = Submodules/nukeupdater
    url = https://github.com/lucasassislar/nukeupdater
    
[submodule "Submodules/TypescriptSyntaxPaste"]
    path = Submodules/TypescriptSyntaxPaste
    url = https://github.com/nhabuiduc/TypescriptSyntaxPaste.git
```

Snailedlt reviewed (commented):
Great work! Almost done!

**Some answer to your earlier comments:**

> - Finalizing README.md content (if we're ok just trimming it to the link sentence + the "Thank You" part, that'd make sense to me)

Fine by me :)

> - Phrasing for Splitscreen Dreams server references in CONTRIBUTING.md (assuming it's moved/removed from the readme)

What we have in CONTRIBUTING.md now is enough imo. Did you think about anything specific?

> Finalizing content/placement of content regarding merging issues with the default branches

This *can* be simplified, but imo it's good enough as it is now, and we can change it later if we want. I made a small suggestion to remove something, after that I think we have something good enough to merge :)

> - I still need to add a reference to the missing dependency to make sure compiling ProtoInput works (I've spent most of today on implementing doc changes so far, but I'm hoping this part doesn't take too long, lol - I'll poke around the ProtoInput page a bit and see if I can get it sorted so that it'll compile with no errors). I could also add a bit about how you can just drop-in the proto input files from the latest release build as a work-around in the troubleshooting section (in case people are getting really stuck on submodule compiling).

How long do you think this will take, and do you need any help with it? If it takes a lot of time, we can merge this PR, and iterate on it afterwards. What do you think?

RangerTJ commented:
>How long do you think this will take, and do you need any help with it? If it takes a lot of time, we can merge this PR, and iterate on it afterwards. What do you think?

Short version is that I'd be fine with merging the PR and iterating later, since I figure this still gives a reasonable starting point, even if there's some known gaps (and I suspect some of the gaps, like compiling Proto Input, are going to always have a certain degree of system-specific troubleshooting involved that may be hard to capture in a general doc). 

_**Long Version**_
I've attached some details of what I've managed to figure out so far, so if anyone has ideas on how to bridge the gaps, I'm totally open to any help on that front. If anyone can get Proto Input and the Nucleus coop to compile and run already without issue, it'd also be super helpful to get a snapshot of what dependencies are installed for Visual Studio / Builder, so I could potentially narrow things down (or at least use them as a frame of reference for a successful compile).

It's taking a bit longer than I expected to get the compiling for Proto Input sorted out, since there's so many potential dependencies, and no docs about it on the Proto Input repo itself. I was _hoping_ it'd just take a day or two but... now I dunno, lol. It's been a bit of random chance as to whether any given dependency changes anything, and there's still the wildcard of a potentially different dev environment than what the original docs were written for being a factor. I have had a little luck today, but I still appear to have one (or more) missing dependencies. Mikou27 mentioned a trial-and-error approach to getting it to work that required so many dependency installs that it's hard to say what the exact requirements are, unfortunately. It sounds like it compiles fine if you basically install _every_ potential dependency, but... there's a _lot_ of them, lol. That said, there's also a known work-around that could be used in the interim, or as the default path (with the caveat that I'd like to double-check with Mikou27 that I'm not missing any other dependencies/files with the current work-around approach).

What I'm inclined to do is modify the compiling instructions for Proto Input to be a branched path, depending on whether you want to work with Proto Input code (which is debatably outside the scope of Nucleus Coop itself anyways). 

**_Workaround Details_**
A workaround exists, in that you can just grab the post-compiled Proto Input files from the most recent Nucleus Coop release, drop those in the root folder, and Nucleus Coop should compile/run fine after that. Mikou27 had mentioned that if you don't plan on working with Proto Input, that's a viable approach. Given contributors to Nucleus Coop are probably going to focus on Nucleus rather than Proto Input, it might even make sense to encourage grabbing the required files from the latest release build as the suggested path (since I'd imagine working with Proto Input would be done through the Proto Input repo anyways).

I've committed a proposed change that has the work-around as the "default" path, if you want to take a look at that. If this seems a bit messy since it's still got some uncertain elements as well, we could potentially just go back to just saying to compile ProtoInput instructions at a high level and mention they may need to ping the developer channel if they run into any issues, or something along those lines (and maybe move the work-around to the bottom section with troubleshooting, etc).

Given how messy the compiling Proto Input is becoming, the more I poke around with it, maybe using the workaround as the suggested starter path makes sense?  I've gotten it to where the only compile errors are "FILE_INFO_BY_HANDLE_CLASS" undefined syntax errors, after installing ATL build tools dependencies and manually editing ProtoInput.sln to fix some stuff that appeared to be compiling out of order. I'm having a rougher time trying to find the source of the syntax error, unfortunately. I'm _reasonably_ sure I have the theoretical dependency installed, at least on paper, so it might be a version issue possibly? Or some permutation of settings/dependencies that results in things installing out of order / triggering syntax errors? Installing everything C++ takes up a _lot_ of space, so I'm kinda hesitant to do that, lol. Assuming the FILE_INFO_BY_HANDLE_CLASS solution fixes the last of the remaining compiling errors, that's the only blocker left to fully documenting how to successfully compile Proto Input (though this assumes there isn't a new error after addressing this, lol).

I've drafted a version of the doc that presents both options, for the time being, so maybe this is a good starting point that can be updated later, and conversation in the Dev channel can fill in the gaps in the meantime?

_**@Mikou27 - Question about the Workaround Approach**_
I noticed that, upon closer inspection, compiling the current build from the master branch relative to the last release results in a few less things showing up in the "utils" folder when I batch build, so I'm not sure if those are things that would have been built from Proto Input? Or just stuff manually added for the release itself? I'm not sure if this is an issue, or normal behavior, in case you have any thoughts on it.

Specifically, the following folders under "util" are all in the release build, but I'm not seeing them when I build from the Nucleus Coop .sln file + drop in the missing DLL's from Proto Input: 
- backup
- dinput8.blocker
- FlawlessWidescreen
- NemirtingasEpicEmu
- NemirtingasGalaxyEmu
- Steam Stub DRM Patcher
- User

I got some crashing behavior once the 2nd game window opened, but it looks like that was just due to the backup folder not existing in "utils" at the time. Creating the folder in release\utils\ directory stopped the crash, so I'm not sure if these folder not showing up in "utils" is still an issue or not. I see them under \splitscreenme-nucleus\Master\NucleusGaming\Tools\, so it seems like it would make sense for them to be built, but it doesn't look like these are actually related to the Proto Input compile issue (since they seem to be things covered under the Nucleus.Gaming tools section). It looks like I might be having a separate issue with Nucleus Coop itself compiling, if these _should_ compile with the rest of the program, but don't. Unless it's a trickle-down effect of Proto Input files not starting off where expected maybe? I noticed lack of Proto Input didn't throw any errors when compiling Nucleus Coop, so maybe it's stealth trickle-down effects from that?

Assuming the above folders in "utils" _aren't_ related to Proto Input, we could probably suggest the "workaround approach" without issue, I think. (Long term it would definitely be good to document how to successfully compile Proto Input so there's better continuity of knowledge on the topic, but that could be addressed separately).

RangerTJ committed:
Update README.md

Trimmed readme based on content migrating to CONTRIBUTING.md

RangerTJ committed:
Update CONTRIBUTING.md

typo fixed

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Update CONTRIBUTING.md

deleted old draft text that refused to die

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Update CONTRIBUTING.md

code highlight cleanup

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Update CONTRIBUTING.md

Workaround path added

RangerTJ committed:
Update CONTRIBUTING.md

Option 1 new line

RangerTJ committed:
Update CONTRIBUTING.md

Polish pass

RangerTJ committed:
Update CONTRIBUTING.md

fixed indent weirdness

RangerTJ committed:
Update CONTRIBUTING.md

fixed for real this time

RangerTJ committed:
Update CONTRIBUTING.md

Fixed a run-on line

RangerTJ commented:
Update: Thanks to help from Mikou27, I got Proto Input to compile! It was kind of a messy process, which was likely related to loading it in VS2022 (and I think some of my earlier issues were from upgrading the project to v143 rather than telling it not to), so I think I may distill it down to a "here's things to check" / be aware of, since it seems very sensitive to your dev environment.

RangerTJ committed:
Update CONTRIBUTING.md

Successful compiling tips added

RangerTJ commented:
Just implemented the additions to the Proto Input compile article and put a good bit about troubleshooting compiling it down in the tips section. I did verify that this didn't make a difference in the folder discrepancy mentioned above, so I think we should be good leaving the work-around as a suggested approach for people that don't plan to mess with proto input source code.

If it turns out that the compile instructions for Nucleus Coop are good, I think all that's left is just a final polish pass on the most recent round of edits. I'm working with Mikou27 regarding release/master branch compile discrepancy in the utils folder. I will update here once I know whether there's a gap I need to address (and if there is, we'll want to update the compile guide for nucleus to include any missing bits).

Update: I think the discrepancy might just be due to the current version of NucleusCoop.sln file on the repo being out of sync with the actual "latest" version used for development (since the repo version that I cloned from shows it was last updated ~2 years ago), so I don't think this is actually an issue with the compiling steps, just a result of compiling with an older .sln file.

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -76,19 +76,44 @@ These will be taken care of as part of the setup git commands, but for reference
-    Close the Proto Input project out when you're done.
+4. Install (or Build) Proto Input
+  
+   **Option 1 (Recomended):** Use Post-Compiled Proto Input Files
```
```suggestion
   **Option 1 (Recommended):** Use Post-Compiled Proto Input Files

```

RangerTJ replied:
Yeah, sorry about the comment wordiness, that is... an ongoing struggle for me, heh. But yeah, everything is basically resolved, I think.

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -76,19 +76,44 @@ These will be taken care of as part of the setup git commands, but for reference
+  
+   **Option 1 (Recomended):** Use Post-Compiled Proto Input Files
+   
+   This is the simplest approach to handling Proto Input file dependencies. If you have no plans to work with the Proto Input source code directly, this is the recomended approach, at is much simpler and has less potential for complications.
```
```suggestion
   This is the simplest approach to handling Proto Input file dependencies. If you have no plans to work with the Proto Input source code directly, this is the recomended approach, as is much simpler and has less potential for complications.

```

RangerTJ replied:
I think I might simplify this even further with a follow-up change, since it was bordering on a run-on sentence. Something like "...this is the recommended approach. It is much simpler and has less potential for complications."

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -76,19 +76,44 @@ These will be taken care of as part of the setup git commands, but for reference
-When you are ready to test any changes you make, compile **Nucleus Coop** by opening its solution file and running a batch build on it within Visual Studio ("Master\NucleusCoop.sln"). Similar to Proto Input, on the Batch Build menu, select all the x86 and x64 options for the build you would like to compile. 
+When you are ready to test any changes you make, compile **Nucleus Coop** by opening its solution file and running a batch build on it within Visual Studio ("Master\NucleusCoop.sln"). 
+
+**Make sure that your project path contains no folders with spaces in their name, or Nucleus Coop may not compile correctly.**
```
very good note! I suggest we make it stand out a bit more by highlighting it as important
```suggestion
> [!IMPORTANT]  
> Make sure that your project path contains no folders with spaces in their name, or Nucleus Coop may not compile correctly.
```

This will look like this:


> [!IMPORTANT]
> Make sure that your project path contains no folders with spaces in their name, or Nucleus Coop may not compile correctly.

Alternatively you could use this:

```suggestion
> [!CAUTION]
> Your project path may not contain any folders with spaces in their name, or Nucleus Coop may not compile correctly.
```

> [!CAUTION]
> Your project path may not contain any folders with spaces in their name, or Nucleus Coop may not compile correctly.

This is a pretty new feature, so in case you want to read more about it, you can do that here: https://github.com/orgs/community/discussions/16925#discussioncomment-9823370

RangerTJ replied:
I was really torn on this one... but I like how "important" flags it without triggering my "something is broken" response like the bright red caution one does when I read it. I'm leaning towards a follow-up commit that blends some of the suggested phrasing in the "caution" suggestion with it too, because I think the stronger tone is a bit clearer (though I think I'll slightly edit it to "Your project path _should_ not contain..."

Snailedlt commented on CONTRIBUTING.md:
````diff
@@ -170,6 +193,13 @@ Even if some of these dependencies are not in active use, the git "update" comma
 ```
 
+#### Issues Compiling Proto Input
+Proto Input appears to be very sensitive to its development environment. If you are opening ProtoInput.sln in Visual Studio 2022, you may be prompted to uprade the project, but doing this might complicate things. If you cannot get it to compile, try opening a fresh version of ProtoInput.sln (so you see the ugprade prompt again) and make sure that all options are set to "No Upgrade". Upgrading can ...
````
```suggestion
Proto Input appears to be very sensitive to its development environment. If you are opening ProtoInput.sln in Visual Studio 2022, you may be prompted to upgrade the project, but doing this might complicate things. If you cannot get it to compile, try opening a fresh version of ProtoInput.sln (so you see the upgrade prompt again) and make sure that all options are set to "No Upgrade". Upgrading can potentially lead to syntax errors in the compiling process. If you see a `FILE_INFO_BY_HANDLE_CLASS` syntax error pop up numerous times during compiling, this is a likely culprit.
```

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -170,6 +193,13 @@ Even if some of these dependencies are not in active use, the git "update" comma
+#### Issues Compiling Proto Input
+Proto Input appears to be very sensitive to its development environment. If you are opening ProtoInput.sln in Visual Studio 2022, you may be prompted to uprade the project, but doing this might complicate things. If you cannot get it to compile, try opening a fresh version of ProtoInput.sln (so you see the ugprade prompt again) and make sure that all options are set to "No Upgrade". Upgrading can ...
+
+If you run into issues compiling ProtoInput.sln, there is a good chance that you are missing some dependencies or add-ons required by Visual Studio or Visual Studrio Build Tools, and you will need to identify and install them. You may also refer to the [Compilation](https://nucleus-coop.github.io/docs/compilation/) article for a more detailed example (with screenshots) of creating builds from the ...
```
```suggestion
If you run into issues compiling ProtoInput.sln, there is a good chance that you are missing some dependencies or add-ons required by Visual Studio or Visual Studio Build Tools, and you will need to identify and install them. You may also refer to the [Compilation](https://nucleus-coop.github.io/docs/compilation/) article for a more detailed example (with screenshots) of creating builds from the source code. Troubleshooting may also involve fixing ProtoInput.sln project build order dependencies, if sub-componenets appear to be building out of the intended order on your setup. In addition to basic C++ dependencies, you will also need to install C++ ATL for build tools (x86/64), at the minimum, and possibly v142 C++ dependencies.
```

Snailedlt commented on CONTRIBUTING.md:
```diff
@@ -170,6 +193,13 @@ Even if some of these dependencies are not in active use, the git "update" comma
+
+If you run into issues compiling ProtoInput.sln, there is a good chance that you are missing some dependencies or add-ons required by Visual Studio or Visual Studrio Build Tools, and you will need to identify and install them. You may also refer to the [Compilation](https://nucleus-coop.github.io/docs/compilation/) article for a more detailed example (with screenshots) of creating builds from the ...
+
+You may also need to update the compiling order. In some dev environments, it appears that the default project compile order may result in certain subcomponents trying to compile before their dependent libraries are finished coding. Keep an eye on output errors from the compile process and note any errors that occur because a file cannot be opened. Make sure that the sub-component that failed to  ...
```
```suggestion
You may also need to update the compiling order. In some dev environments, it appears that the default project compile order may result in certain subcomponents trying to compile before their dependent libraries are finished coding. Keep an eye on output errors from the compile process and note any errors that occur because a file cannot be opened. Make sure that the sub-component that failed to compile has the project that contains the file that failed to open as a dependency. To set dependencies, right-click the ProtoInput solution in the solution explorer section of Visual Studio and select the "Project Dependencies..." option. From there, simply use the drop-down box to select a subcomponent and then check the boxes next to any known, missing dependencies (as seen from output errors). Submodules with known issues in certain setups include _ProtoInputInjectorProxy_ (requires EasyHookDLL) and _ProtoInputUtilStatic_ (requies Blackbone). Check these submodules first and ensure they have their dependencies checked if you have any issues. Otherwise use output errors to help troubleshoot any other dependency issues - a file failing to open is a good indicator that it wasn't installed yet (and is thus is a dependency that should have been installed earlier). 
```

RangerTJ replied:
Just caught another typo in this that I'll fix in a follow-up, towards the end I snuck in an extra "is".

 _"...(and is thus **is** a dependency that should have been installed earlier). "_

Snailedlt reviewed (dismissed):
I only skim read through your comments (they're so long!), but I took it you'd solved most of the issues, and wanted to iterate on the parts missing, which I agree is a good solution.

Now there's just some small typos left, and we're all good, great work! 🥇

RangerTJ committed:
Update CONTRIBUTING.md

typo fixed

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Update CONTRIBUTING.md

typo fixed

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Update CONTRIBUTING.md

Better highlighting of the syntax errors on compile

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Update CONTRIBUTING.md

typo fixed

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Update CONTRIBUTING.md

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Update CONTRIBUTING.md

Better formatting for important compile note

Co-authored-by: Jørgen Kalsnes Hagen <43886029+Snailedlt@users.noreply.github.com>

RangerTJ committed:
Update CONTRIBUTING.md

Polish tweaks to phrasing for a few lines

RangerTJ committed:
Update CONTRIBUTING.md

Typo cleanup pass based on running the current version through MS word spelling/grammar check

RangerTJ commented:
Added a final(?) round of polish edits based on the last round of feedback for a few things that I thought would help beyond the suggested commits. Most of the additional changes were slight alterations that I thought might help with language clarity a bit. Mostly it was adapting the "caution" suggestion tone to the "important" suggestion version for the compile path note, and breaking a couple long sentences into multiple sentences. This should hopefully help with readability. 

I also took the final product of those edits and then ran it through MS word to do a spelling/grammar check, and caught a few more things that slipped through the cracks.

Snailedlt commented:
@Mikou27 Please look over the technical details again to make sure it still looks good :)

Mikou27 commented:
Awesome work! Thank you very much for this huge contribution to the project!

Snailedlt commented:
Yes, great work @RangerTJ!
